### PR TITLE
ZBUG-3364:Added test case to validate service.PERM_DENIED is not coming when sending email as distribution list and using persona.

### PR DESCRIPTION
**Pull request by pallavikhairnar21**
ZBUG-3364:Added test case to validate error-service.PERM_DENIED is not coming when sending email as distribution list and using persona.
Test case 1: Verify sendAsDistList DL rights and no service.PERM_DENIED error when sending email as distribution list and using persona
Steps:
                        1. Login into Admin.
			2. Grant Permission "sendAsDistList" to Account1 on Distribution list.
			3. Login into Account1.
			4. Send DiscoverRightsRequest to discover grants granted.
			5. Send email with From: set to DL and Verify message details using GetMsgRequest
			6. Create a persona for DL 
			7. Send an email using persona and verify service.PERM_DENIED error is not encountered
			8. Verify message details using GetMsgRequest
Test case 2: Verify sendOnBehalfOfDistList DL rights and no service.PERM_DENIED error when sending email as distribution list and using persona
Steps:
                        1. Login into Admin.
			2. Grant Permission "sendOnBehalfOfDistList" to Account1 on Distribution list.
			3. Login into Account1.
			4. Send DiscoverRightsRequest to discover grants granted.
			5. Send email with From: set to DL and Verify message details using GetMsgRequest
			6. Create a persona for DL 
			7. Send an email using persona and verify service.PERM_DENIED error is not encountered
			8. Verify message details using GetMsgRequest
I will add additional test cases in next automation ticket..

Test automation result:
XML: C:\soap\zm-soap-harness\data\soapvalidator\Mail\Persona\SendMailUsingDistributionListPersona.xml

-----------------------------REQUEST PERFORMANCE SUMMARY---------------------------------

Request Name                              ExecutionTime(msec)  ExecutionCount  AvgExecutionTime(msec)

AuthRequest                                       10237          7           1462
CreateAccountRequest                              14214          3           4738
DiscoverRightsRequest                              3172          2           1586
SearchRequest                                      3272          2           1636
PingRequest                                        1941          1           1941
GrantRightRequest                                  3069          2           1534
CreateDistributionListRequest                      3155          2           1577
AddDistributionListMemberRequest                   2907          2           1453
CreateIdentityRequest                              2964          2           1482
GetMsgRequest                                      9037          6           1506
SendMsgRequest                                     5533          4           1383

-------------------SUMMARY---------------------

PASS 0001   1/1  1941ms PingRequest                 	Ping
PASS 0002   1/1  1423ms AuthRequest                 	acct1_setup
PASS 0003   2/2 10330ms CreateAccountRequest        	acct1_setup
PASS 0004   2/2  1980ms CreateAccountRequest        	acct1_setup
PASS 0005   2/2  1904ms CreateAccountRequest        	acct1_setup
PASS 0006   1/1  1623ms CreateDistributionListRequest	acct1_setup
PASS 0007   1/1  1244ms AddDistributionListMemberRequest	acct1_setup
PASS 0008   1/1  1532ms CreateDistributionListRequest	acct1_setup
PASS 0009   1/1  1663ms AddDistributionListMemberRequest	acct1_setup
PASS 0010   1/1  1242ms AuthRequest                 	sendMailAsDL
PASS 0011   1/1  1433ms GrantRightRequest           	sendMailAsDL
PASS 0012   2/2  1333ms AuthRequest                 	sendMailAsDL
PASS 0013   2/2  1533ms DiscoverRightsRequest       	sendMailAsDL
PASS 0014   2/2  1436ms SendMsgRequest              	sendMailAsDL
PASS 0015   4/4  1539ms GetMsgRequest               	sendMailAsDL
PASS 0016   3/3  1325ms CreateIdentityRequest       	sendMailAsDL
PASS 0017   2/2  1427ms SendMsgRequest              	sendMailAsDL
PASS 0018   5/5  1440ms GetMsgRequest               	sendMailAsDL
PASS 0019   2/2  1637ms AuthRequest                 	sendMailAsDL
PASS 0020   2/2  1636ms SearchRequest               	sendMailAsDL
PASS 0021   5/5  1635ms GetMsgRequest               	sendMailAsDL
PASS 0022   1/1  1636ms AuthRequest                 	sendMailOnBehalfOfDL
PASS 0023   1/1  1636ms GrantRightRequest           	sendMailOnBehalfOfDL
PASS 0024   2/2  1330ms AuthRequest                 	sendMailOnBehalfOfDL
PASS 0025   2/2  1639ms DiscoverRightsRequest       	sendMailOnBehalfOfDL
PASS 0026   2/2  1325ms SendMsgRequest              	sendMailOnBehalfOfDL
PASS 0027   4/4  1448ms GetMsgRequest               	sendMailOnBehalfOfDL
PASS 0028   3/3  1639ms CreateIdentityRequest       	sendMailOnBehalfOfDL
PASS 0029   2/2  1345ms SendMsgRequest              	sendMailOnBehalfOfDL
PASS 0030   5/5  1646ms GetMsgRequest               	sendMailOnBehalfOfDL
PASS 0031   2/2  1636ms AuthRequest                 	sendMailOnBehalfOfDL
PASS 0032   2/2  1636ms SearchRequest               	sendMailOnBehalfOfDL
PASS 0033   5/5  1329ms GetMsgRequest               	sendMailOnBehalfOfDL

script_parsable:	2	0

TEST CASE Results: 2(PASS) - 0(Fail)

REQUEST/RESPONSE Results: 33(PASS) - 0(Fail)

### CODE REVIEW

**Review by deepakzimb — APPROVED** [review]
LGTM

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+				</m>
+			</SendMsgRequest>
+		</t:request>
+		<t:response>
+			<t:select path="//zimbra:Code" match="service.PERM_DENIED" emptyset="1"/>
+			<t:select path="//mail:SendMsgResponse/mail:m" attr="id" set="message.id1"/>
+		</t:response>
+	</t:test>
+	
+	<!-- Verify message details using GetMsgRequest -->
```
Comment (jiteshsojitra) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:278]: indentation

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+
+	
+	<!-- Verify message details using GetMsgRequest -->
+    <t:test id="GetMsgRequest">
+        <t:request>
+            <GetMsgRequest xmlns="urn:zimbraMail">
+                <m id="${message.id2}"/>
+            </GetMsgRequest>
+        </t:request>
+       <t:response>
```
Comment (jiteshsojitra) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:341]: indentation

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml [hunk 3]
```diff
@@ -0,0 +1,516 @@
+<t:tests xmlns:t="urn:zimbraTestHarness">
+
+<t:property name="test_account1.name" value="test1.${TIME}.${COUNTER}@${defaultdomain.name}"/>
+<t:property name="test_account2.name" value="test2.${TIME}.${COUNTER}@${defaultdomain.name}"/>
```
Comment (jiteshsojitra) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:4]: is this account used for delegated testcases?

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+<t:property name="test_account3.name" value="test3.${TIME}.${COUNTER}@${defaultdomain.name}"/>
```
Comment (jiteshsojitra) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:5]: is this account used for delegated testcases?

**Comment by jiteshsojitra** [comment]
Bugs/ZBUG-3364.xml

Please stop using adding testcase under Bugs, create proper named file under valid location for e.g. SendMailUsingPersonaDelegation or SendMailUsingPersonaDistributionList something like this the way SOAP harness follows.

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml (hunk 1 above)
Comment (pallavikhairnar21) [reply data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:278]: Corrected indentation

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml (hunk 2 above)
Comment (pallavikhairnar21) [reply data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:341]: Corrected indentation

Comment (jiteshsojitra) [reply data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:341]: entire file has indentation problem, please use tabs everywhere if that is followed.

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml (hunk 3 above)
Comment (pallavikhairnar21) [reply data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:4]: test_account1.name is the account which is part of DL and has sendAsDistList right. I will add few more test cases in the next sprint so keep test_account2.name as well ZCS-13336. If needed I can delete it - test_account2.name.

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml (hunk 4 above)
Comment (jiteshsojitra) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:5]: is this account used for delegated testcases?

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+					<su> ${msg01.subject}</su>
+					<mp ct="text/plain">
+					<content> ${message.content1}</content>
+					</mp>
+				</m>
+			</SendMsgRequest>
+		</t:request>
+		<t:response>
+			<t:select path="//zimbra:Code" match="service.PERM_DENIED" emptyset="1"/>
+			<t:select path="//mail:SendMsgResponse/mail:m" attr="id" set="message.id1"/>
```
Comment (jiteshsojitra) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml:274]: Please verify from headers (who is sending mail / send as).

**Comment by pallavikhairnar21** [comment]
> Bugs/ZBUG-3364.xml
> 
> Please stop using adding testcase under Bugs, create proper named file under valid location for e.g. SendMailUsingPersonaDelegation or SendMailUsingPersonaDistributionList something like this the way SOAP harness follows.

Added Persona folder in Mail and added SendMailUsingDistributionListPersona in it..

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml (hunk 3 above)
Comment (prashantajari) [reply data/soapvalidator/Mail/Bugs/ZBUG-3364.xml]: Lets not create too many test accounts. I think, we can restrict account to account1, account2 (as DL members) and account3 as recipient user.

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+    <t:test id="create_acct1" required="true" depends="admin_login">
+        <t:request>
+            <CreateAccountRequest xmlns="urn:zimbraAdmin">
+                <name>${test_account1.name}</name>
+                <password>${defaultpassword.value}</password>
+            </CreateAccountRequest>
+            </t:request>
+            <t:response>
+                <t:select path="//admin:CreateAccountResponse/admin:account" attr="id"  set="test_account1.id"/>
+                <t:select path='//admin:CreateAccountResponse/admin:account/admin:a[@n="zimbraMailHost"]' set="test_acct1.server"/>
```
Comment (prashantajari) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml]: test_acct1.server and similar values are not used in test and should be removed

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+       <t:response>
+			<t:select path="//mail:GetMsgResponse/mail:m[@id='${message.id2}']">
+            <t:select path="//mail:e[@t='f']" attr="a" match="${test_list1.name}"/>
+            <t:select path="//mail:e[@t='f']" attr="p" match="${persona.name}"/>
+            <t:select path="//mail:e[@t='t']" attr="a" match="${test_account4.name}"/>
+            <t:select path="//mail:su" match="${msg02.subject}"/>
+			</t:select>
+			</t:response>
+	</t:test>
+</t:test_case>
```
Comment (prashantajari) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml]: Please add validation at recipient end to check from/to address and other message details

File: data/soapvalidator/Mail/Bugs/ZBUG-3364.xml [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+       <t:response>
+			<t:select path="//mail:GetMsgResponse/mail:m[@id='${message.id3}']">
+            <t:select path="//mail:e[@t='f']" attr="a" match="${test_list2.name}"/>
+            <t:select path="//mail:e[@t='f']" attr="p" match="${persona02.name}"/>
+            <t:select path="//mail:e[@t='t']" attr="a" match="${test_account4.name}"/>
+            <t:select path="//mail:su" match="${msg03.subject}"/>
+			</t:select>
+			</t:response>
+    </t:test>
+
```
Comment (prashantajari) [inline data/soapvalidator/Mail/Bugs/ZBUG-3364.xml]: Please add validation at recipient end to check from/to address and other message details

**Review by prashantajari — CHANGES_REQUESTED** [review]
Minor changes requested

**Comment by pallavikhairnar21** [comment]
> Minor changes requested

Fixed comments provided by Prashant and Jitest, please check review again

**Comment by jiteshsojitra** [comment]
1. Still there were many indentation issues which was not fixed.
2. There should be only 1 commit with rebased with master/develop.

**Pull request merged**